File: src/main/java/com/heybys/oddments/security/OAuth2UserPrincipal.java
package com.heybys.oddments.security;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.oauth2.core.user.OAuth2User;

import com.heybys.oddments.fooddelivery.domain.user.User;

public class OAuth2UserPrincipal implements OAuth2User, UserDetails {
    private final Long id;
    private final String name;
    private final String email;
    private final String password;
    private final Collection<? extends GrantedAuthority> authorities;
    private final Map<String, Object> attributes;

    public OAuth2UserPrincipal(
            Long id,
            String name,
            String email,
            String password,
            Collection<? extends GrantedAuthority> authorities,
            Map<String, Object> attributes) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.password = password;
        this.authorities = authorities;
        this.attributes = attributes;
    }

    public static OAuth2UserPrincipal create(User user) {
        return OAuth2UserPrincipal.create(user, null);
    }

    public static OAuth2UserPrincipal create(User user, Map<String, Object> attributes) {
        List<GrantedAuthority> authorities = Collections.singletonList(new SimpleGrantedAuthority("ROLE_USER"));

        return new OAuth2UserPrincipal(
                user.getId().longValue(), user.getName(), user.getEmail(), user.getPassword(), authorities, attributes);
    }

    public Long getId() {
        return id;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public Map<String, Object> getAttributes() {
        return attributes;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authorities;
    }

    @Override
    public String getUsername() {
        return email;
    }

    @Override
    public String getPassword() {
        return password;
    }

    @Override
    public boolean isAccountNonExpired() {
        return true;
    }

    @Override
    public boolean isAccountNonLocked() {
        return true;
    }

    @Override
    public boolean isCredentialsNonExpired() {
        return true;
    }

    @Override
    public boolean isEnabled() {
        return true;
    }
}

File: src/main/kotlin/com/heybys/oddments/security/OAuth2UserPrincipal.kt
package com.heybys.oddments.security

import com.heybys.oddments.fooddelivery.domain.user.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.oauth2.core.user.OAuth2User

class OAuth2UserPrincipal(
    private val id: Long,
    private val name: String,
    private val email: String,
    private val password: String,
    private val authorities: Collection<GrantedAuthority>,
    private val attributes: Map<String, Any>?,
) : OAuth2User, UserDetails {

    companion object {
        fun create(user: User): OAuth2UserPrincipal {
            return create(user, null)
        }

        fun create(user: User, attributes: Map<String, Any>?): OAuth2UserPrincipal {
            val authorities = listOf(SimpleGrantedAuthority("ROLE_USER"))

            return OAuth2UserPrincipal(
                user.getId()?.longValue() ?: 0L,
                user.name ?: "",
                user.email ?: "",
                user.password ?: "",
                authorities,
                attributes,
            )
        }
    }

    fun getId(): Long = id

    override fun getName(): String = name

    override fun getAttributes(): Map<String, Any>? = attributes

    override fun getAuthorities(): Collection<GrantedAuthority> = authorities

    override fun getUsername(): String = email

    override fun getPassword(): String = password

    override fun isAccountNonExpired(): Boolean = true

    override fun isAccountNonLocked(): Boolean = true

    override fun isCredentialsNonExpired(): Boolean = true

    override fun isEnabled(): Boolean = true
}
